File: scratch/calculate_metrics.py
import csv
import math
import os

def pearson_correlation(x, y):
    n = len(x)
    if n == 0:
        return 0.0
    mean_x = sum(x) / n
    mean_y = sum(y) / n
    
    num = 0.0
    den_x = 0.0
    den_y = 0.0
    
    for i in range(n):
        diff_x = x[i] - mean_x
        diff_y = y[i] - mean_y
        num += diff_x * diff_y
        den_x += diff_x * diff_x
        den_y += diff_y * diff_y
        
    if den_x == 0.0 or den_y == 0.0:
        return 0.0
    return num / math.sqrt(den_x * den_y)

def quadratic_weighted_kappa(x, y, min_val=0, max_val=100):
    n = len(x)
    if n == 0:
        return 0.0
        
    rounded_x = [round(v) for v in x]
    rounded_y = [round(v) for v in y]
    
    classes = list(range(min_val, max_val + 1))
    num_classes = len(classes)
    class_index = {c: idx for idx, c in enumerate(classes)}
    
    # Observational matrix O
    O = [[0] * num_classes for _ in range(num_classes)]
    hist_x = [0] * num_classes
    hist_y = [0] * num_classes
    
    for i in range(n):
        val_x = rounded_x[i]
        val_y = rounded_y[i]
        idx_x = class_index.get(val_x)
        idx_y = class_index.get(val_y)
        if idx_x is not None and idx_y is not None:
            O[idx_x][idx_y] += 1
            hist_x[idx_x] += 1
            hist_y[idx_y] += 1
            
    # Weights matrix W
    W = [[0.0] * num_classes for _ in range(num_classes)]
    for i in range(num_classes):
        for j in range(num_classes):
            diff = classes[i] - classes[j]
            W[i][j] = (diff * diff) / ((num_classes - 1) ** 2) if num_classes > 1 else 0.0
            
    # Expected matrix E
    E = [[0.0] * num_classes for _ in range(num_classes)]
    for i in range(num_classes):
        for j in range(num_classes):
            E[i][j] = (hist_x[i] * hist_y[j]) / n
            
    # Compute agreement and expected disagreement
    num = 0.0
    den = 0.0
    for i in range(num_classes):
        for j in range(num_classes):
            num += W[i][j] * O[i][j]
            den += W[i][j] * E[i][j]
            
    if den == 0.0:
        return 0.0
    return 1.0 - (num / den)

def generate_visualization(x_ai, y_dosen, output_path):
    try:
        import matplotlib.pyplot as plt
        
        plt.figure(figsize=(8, 6), dpi=300)
        
        # Plot scatter points
        plt.scatter(x_ai, y_dosen, color='#4f46e5', alpha=0.7, edgecolors='none', s=60, label='Mahasiswa (N=33)')
        
        # Perfect agreement reference diagonal line
        plt.plot([0, 100], [0, 100], color='#10b981', linestyle='--', linewidth=1.5, label='Garis Keselarasan Sempurna (y=x)')
        
        # Set titles & labels
        plt.title('Perbandingan Skor Otomatis AI vs Skor Manual Dosen\n(Smart Assistant Lecturer)', fontsize=12, fontweight='bold', pad=15)
        plt.xlabel('Skor Penilaian AI', fontsize=10)
        plt.ylabel('Skor Penilaian Dosen', fontsize=10)
        
        plt.xlim(0, 105)
        plt.ylim(0, 105)
        plt.grid(True, linestyle=':', alpha=0.6)
        plt.legend(loc='lower right', fontsize=9)
        
        plt.tight_layout()
        plt.savefig(output_path)
        plt.close()
        print(f"Grafik visualisasi sebaran nilai berhasil disimpan di: {output_path}")
        
    except ImportError:
        print("\n[INFO] Pustaka 'matplotlib' tidak terpasang.")
        print("Untuk menghasilkan grafik visualisasi, silakan jalankan:")
        print("  pip install matplotlib")

def main():
    csv_path = os.path.join(os.path.dirname(__file__), "../docs/IF23A_cleaned.csv")
    img_output_path = os.path.join(os.path.dirname(__file__), "../docs/sebaran_nilai.png")
    
    if not os.path.exists(csv_path):
        print(f"File CSV tidak ditemukan di: {csv_path}")
        return

    x_ai = []
    y_dosen = []
    
    with open(csv_path, mode="r", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            x_ai.append(float(row["skor_ai"]))
            y_dosen.append(float(row["skor_dosen"]))
            
    n = len(x_ai)
    print(f"Membaca {n} data mahasiswa dari {csv_path}...")
    
    # Calculate metrics
    r = pearson_correlation(x_ai, y_dosen)
    qwk100 = quadratic_weighted_kappa(x_ai, y_dosen, 0, 100)
    
    # Binned 0-10 scale
    x_ai_10 = [v / 10.0 for v in x_ai]
    y_dosen_10 = [v / 10.0 for v in y_dosen]
    qwk10 = quadratic_weighted_kappa(x_ai_10, y_dosen_10, 0, 10)
    
    print("\n=== HASIL EVALUASI STATISTIK (LOKAL) ===")
    print(f"1. Pearson Correlation Coefficient (r)                  : {r:.4f}")
    print(f"2. Quadratic Weighted Kappa (QWK) Skala 100 (Integer)  : {qwk100:.4f}")
    print(f"3. Quadratic Weighted Kappa (QWK) Skala 10  (Binned)   : {qwk10:.4f}")
    print("=========================================\n")
    
    # Try to generate plot
    generate_visualization(x_ai, y_dosen, img_output_path)

if __name__ == "__main__":
    main()

File: scratch/calculate_metrics.js
const { createClient } = require("@supabase/supabase-js");
const fs = require("fs");
const path = require("path");

// Load environment variables
const envContent = fs.readFileSync(path.join(__dirname, "..", ".env"), "utf8");
const envVars = {};
envContent.split("\n").forEach(line => {
  const parts = line.split("=");
  if (parts.length >= 2) {
    const key = parts[0].trim();
    const value = parts.slice(1).join("=").trim().replace(/^['"]|['"]$/g, '');
    envVars[key] = value;
  }
});

const supabase = createClient(
  envVars.NEXT_PUBLIC_SUPABASE_URL,
  envVars.NEXT_PUBLIC_SUPABASE_ANON_KEY
);

// Pearson correlation coefficient calculation
function pearsonCorrelation(x, y) {
  const n = x.length;
  if (n === 0) return 0;
  
  const sumX = x.reduce((a, b) => a + b, 0);
  const sumY = y.reduce((a, b) => a + b, 0);
  
  const meanX = sumX / n;
  const meanY = sumY / n;
  
  let num = 0;
  let denX = 0;
  let denY = 0;
  
  for (let i = 0; i < n; i++) {
    const diffX = x[i] - meanX;
    const diffY = y[i] - meanY;
    num += diffX * diffY;
    denX += diffX * diffX;
    denY += diffY * diffY;
  }
  
  if (denX === 0 || denY === 0) return 0;
  return num / Math.sqrt(denX * denY);
}

// Quadratic Weighted Kappa (QWK) calculation
function quadraticWeightedKappa(x, y, minVal = 0, maxVal = 100) {
  const n = x.length;
  if (n === 0) return 0;

  const roundedX = x.map(v => Math.round(v));
  const roundedY = y.map(v => Math.round(v));
  
  // Define rating classes
  const classes = [];
  for (let c = minVal; c <= maxVal; c++) {
    classes.push(c);
  }
  const numClasses = classes.length;
  const classIndex = {};
  classes.forEach((c, idx) => {
    classIndex[c] = idx;
  });

  // Initialize matrices
  const O = Array(numClasses).fill(0).map(() => Array(numClasses).fill(0));
  const W = Array(numClasses).fill(0).map(() => Array(numClasses).fill(0));
  
  // Calculate histogram counts for expected matrix
  const histX = Array(numClasses).fill(0);
  const histY = Array(numClasses).fill(0);
  
  for (let i = 0; i < n; i++) {
    const idxX = classIndex[roundedX[i]];
    const idxY = classIndex[roundedY[i]];
    
    if (idxX !== undefined && idxY !== undefined) {
      O[idxX][idxY]++;
      histX[idxX]++;
      histY[idxY]++;
    }
  }
  
  // Calculate weights matrix
  for (let i = 0; i < numClasses; i++) {
    for (let j = 0; j < numClasses; j++) {
      const diff = classes[i] - classes[j];
      W[i][j] = (diff * diff) / ((numClasses - 1) * (numClasses - 1));
    }
  }
  
  // Calculate expected matrix E
  const E = Array(numClasses).fill(0).map(() => Array(numClasses).fill(0));
  for (let i = 0; i < numClasses; i++) {
    for (let j = 0; j < numClasses; j++) {
      E[i][j] = (histX[i] * histY[j]) / n;
    }
  }
  
  // Calculate QWK score
  let num = 0;
  let den = 0;
  for (let i = 0; i < numClasses; i++) {
    for (let j = 0; j < numClasses; j++) {
      num += W[i][j] * O[i][j];
      den += W[i][j] * E[i][j];
    }
  }
  
  if (den === 0) return 0;
  return 1 - (num / den);
}

async function compute() {
  console.log("Fetching live submissions from Supabase database...");
  const { data: submissions, error } = await supabase
    .from("submissions")
    .select("nim, student_name, ai_score, final_score, status")
    .order("nim", { ascending: true });

  if (error) {
    console.error("Database error:", error);
    process.exit(1);
  }

  console.log(`Successfully fetched ${submissions.length} submissions.`);

  const dataset = submissions.map(sub => {
    return {
      nim: sub.nim,
      name: sub.student_name,
      aiScore: parseFloat(sub.ai_score || 0),
      finalScore: parseFloat(sub.final_score || 0),
      status: sub.status
    };
  });

  const aiScoresList = dataset.map(d => d.aiScore);
  const finalScoresList = dataset.map(d => d.finalScore);

  console.log("\n=== DAFTAR NILAI LIVE MAHASISWA ===");
  console.log("No | NIM | Nama Mahasiswa | Skor AI | Skor Akhir (Dosen) | Status");
  console.log("----------------------------------------------------------------------");
  dataset.forEach((d, idx) => {
    console.log(
      `${String(idx + 1).padStart(2, ' ')} | ${d.nim} | ${d.name.padEnd(30, ' ')} | ${String(d.aiScore.toFixed(1)).padStart(7, ' ')} | ${String(d.finalScore.toFixed(1)).padStart(18, ' ')} | ${d.status}`
    );
  });

  // Calculate Pearson correlation between AI score and Final Score
  const pearsonVal = pearsonCorrelation(aiScoresList, finalScoresList);
  
  // Calculate QWK on scale 0-100 (nearest rounded integer)
  const qwk100 = quadraticWeightedKappa(aiScoresList, finalScoresList, 0, 100);
  
  // Calculate QWK on scale 0-10 (binned by dividing by 10)
  const binnedAIList = aiScoresList.map(s => s / 10);
  const binnedFinalList = finalScoresList.map(s => s / 10);
  const qwk10 = quadraticWeightedKappa(binnedAIList, binnedFinalList, 0, 10);

  console.log("\n=== HASIL EVALUASI STATISTIK DARI DATABASE ===");
  console.log(`1. Koefisien Korelasi Pearson (r): ${pearsonVal.toFixed(4)}`);
  printInterpretation(pearsonVal);
  console.log(`2. Quadratic Weighted Kappa (QWK) Skala 100: ${qwk100.toFixed(4)}`);
  console.log(`3. Quadratic Weighted Kappa (QWK) Skala 10 (Binned): ${qwk10.toFixed(4)}`);
  console.log("=================================================");
  
  // Save dataset to JSON
  fs.writeFileSync(
    path.join(__dirname, "evaluation_data.json"),
    JSON.stringify({ dataset, pearson: pearsonVal, qwk100, qwk10 }, null, 2),
    "utf8"
  );
  console.log("\nSaved dataset copy to scratch/evaluation_data.json");
}

function printInterpretation(r) {
  let strength = "";
  if (Math.abs(r) < 0.2) strength = "Sangat Lemah / Hampir Tidak Ada Korelasi";
  else if (Math.abs(r) < 0.4) strength = "Korelasi Lemah";
  else if (Math.abs(r) < 0.6) strength = "Korelasi Sedang";
  else if (Math.abs(r) < 0.8) strength = "Korelasi Kuat";
  else strength = "Korelasi Sangat Kuat";
  console.log(`   └─ Interpretasi: Keandalan Hubungan ${strength}`);
}

compute();
